### ISSUE TITLE
Implement subList() on primitive lists

### ISSUE THREAD

**Issue by vmzakharov** [eclipse/eclipse-collections#1053]
Currently `subList()` on primitive lists is stubbed with
```
throw new UnsupportedOperationException("subList not yet implemented!");
```
The new behavior should be consistent with the behaviors currently defined in `FastList` for the mutable lists and `AbstractImmutableList` for the immutable ones.

**Comment by gaurangkudale** [comment]
Hi @vmzakharov,
Can I work on this issue? if yes can please direct me to the code line

**Comment by sneznaj** [comment]
i am already working on this

**Comment by sneznaj** [comment]
i am already working on this

**Comment by gaurangkudale** [comment]
Hi, @sneznaj Do you need any help? we can collaborate

**Comment by sneznaj** [comment]
@vmzakharov FastList does not contain a sublist method or class

**Comment by hugo-we** [comment]
any progress?

**Comment by knowledge-graphlet** [comment]
```
public ImmutableIntList subList(int fromIndex, int toIndex) {
    if (fromIndex < 0 || toIndex > this.size() || fromIndex > toIndex) {
        throw new IndexOutOfBoundsException("From: " + fromIndex + ", To: " + toIndex + ", Size: " + this.size());
    }

    if (fromIndex == toIndex) {
        return IntLists.immutable.empty();
    }
    
    if (fromIndex == 0 && toIndex == this.size()) {
        return this;
    }

    MutableIntList subList = IntLists.mutable.withInitialCapacity(toIndex - fromIndex);
    for (int i = fromIndex; i < toIndex; i++) {
        subList.add(this.get(i));
    }
    return subList.toImmutable();
}

```

Is this an acceptable implementation?

**Comment by mohrezaei** [comment]
No, it's not. Sub-lists in java are not copies.

**Comment by knowledge-graphlet** [comment]
For mutable lists, I can see the distinction... Presuming you want the sublist to be a projection of the "backing" list, so that updates are reflected in the sublist. 

But if the list is immutable, and there is no object identity in the values, and there is no ability to update elements (so no notion of updates from the "backing" list). So, I'm not sure what you would be trying to accomplish that an immutable list that copies the range doesn't provide.  It seems you could have a pretty simple set of implementations for the primitive immutable types.

**Comment by mohrezaei** [comment]
For immutable lists, it's even more important, because the operation is O(1) and there is no memory overhead. FYI, in languages that are immutable by design (e.g. Lisp), all implementations of sub lists (`car`, `cdr`, etc) are by reference.

**Comment by knowledge-graphlet** [comment]
Fair enough, I can see that for very large immutable lists, there is a memory conservation goal that is independent of being a value-type object.
